### PR TITLE
improve error handling and logging in s3-storeuploader

### PR DESCRIPTION
**Pull request by xrstf**
**What this PR does / why we need it**:
This replaces glog with zap in the s3-storeuploader and generally improves its error handling. The most important fix is that bad endpoints (bad URL, HTTP failures, network failures) are no longer hidden behind a bogus "object name cannot be empty" error, but rather displayed to the user:

```
./s3-storeuploader delete-all --endpoint=foo --prefix=foo
2019-08-02T22:14:46.143+0200    fatal   s3-storeuploader/main.go:153    Failed to run command   {"err": "Get http://foo/kubermatic-backups/?location=: dial tcp: lookup foo: no such host"}
```

Likewise, empty prefixes are no longer allowed.

I made the two new logging CLI flags global because otherwise I could not access them in my global Before() handler. Alternatively we could (maybe?) use a Before() handler for each action, so that the CLI flags can be per-command. Given urfave's parsing, users now **must** put the logging flags before the subcommand name: `./s3-storeuploader --log-debug delete-all --endpoint=...`.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

File: api/cmd/s3-storeuploader/main.go [hunk 1]
```diff
@@ -108,27 +128,53 @@ func main() {
...
+	}()
+
 	err := app.Run(os.Args)
-	if err != nil {
-		log.Fatal(err)
+	// Only log failures when the logger has been setup, otherwise
+	// we know it's been a CLI parsing failure and the cli package
+	// has already output the error and printed the usage hints.
+	if err != nil && logger != nil {
+		logger.Fatalw("Failed to run command", "err", err)
```
Comment (mrIncompetent) [inline api/cmd/s3-storeuploader/main.go]: nit: use `zap.Error(err)` instead. That ensures we have a consistent error key

File: api/pkg/storeuploader/storeuploader.go [hunk 2]
```diff
@@ -21,54 +23,76 @@ const prefixSeparator = "storeuploader"
...
+		if object.Err != nil {
+			return object.Err
+		}
 		existingObjects = append(existingObjects, object)
 	}
 
+	u.logger.Debugw("Done listing bucket", "bucket", bucket, "prefix", prefix, "objects", len(existingObjects))
+
 	for _, object := range u.getObjectsToDelete(existingObjects, revisionsToKeep) {
+		u.logger.Debugw("Removing object", "bucket", bucket, "object", object.Key)
```
Comment (mrIncompetent) [inline api/pkg/storeuploader/storeuploader.go]: Info

File: api/pkg/storeuploader/storeuploader.go [hunk 3]
```diff
@@ -77,17 +101,29 @@ func (u *StoreUploader) DeleteOldBackups(file, bucket, prefix string, revisionsT
...
+		if object.Err != nil {
+			return object.Err
+		}
 		existingObjects = append(existingObjects, object)
 	}
 
+	u.logger.Debugw("Done listing bucket", "bucket", bucket, "prefix", prefix, "objects", len(existingObjects))
+
 	for _, object := range existingObjects {
+		u.logger.Debugw("Removing object", "bucket", bucket, "object", object.Key)
```
Comment (mrIncompetent) [inline api/pkg/storeuploader/storeuploader.go]: Info

File: api/pkg/storeuploader/storeuploader.go [hunk 4]
```diff
@@ -21,54 +23,76 @@ const prefixSeparator = "storeuploader"
...
+// DeleteOldBackups deletes revisions of all files of the given prefix which are older than max-revisions
+func (u *StoreUploader) DeleteOldBackups(bucket, prefix string, revisionsToKeep int) error {
+	if len(prefix) == 0 {
+		return errors.New("prefix cannot be empty")
+	}
+
 	doneCh := make(chan struct{})
 	defer close(doneCh)
 
+	u.logger.Debugw("Listing existing objects", "bucket", bucket, "prefix", prefix)
```
Comment (mrIncompetent) [inline api/pkg/storeuploader/storeuploader.go]: Create a sub logger here with bucket & prefix. That way you avoid setting those fields for the following calls.
This also reduces the chances for typos in keys

Comment (mrIncompetent) [reply api/pkg/storeuploader/storeuploader.go]: Oh and please add `revisions-to-keep`

File: api/pkg/storeuploader/storeuploader.go [hunk 5]
```diff
@@ -77,17 +101,29 @@ func (u *StoreUploader) DeleteOldBackups(file, bucket, prefix string, revisionsT
...
+// DeleteAll deletes all revisions of all files matching the given prefix
+func (u *StoreUploader) DeleteAll(bucket, prefix string) error {
+	if len(prefix) == 0 {
+		return errors.New("prefix cannot be empty")
+	}
+
 	doneCh := make(chan struct{})
 	defer close(doneCh)
 
+	u.logger.Debugw("Listing existing objects", "bucket", bucket, "prefix", prefix)
```
Comment (mrIncompetent) [inline api/pkg/storeuploader/storeuploader.go]: https://github.com/kubermatic/kubermatic/pull/3943/files#r310285769

File: api/pkg/storeuploader/storeuploader.go (hunk 4 above)
Comment (xrstf) [reply api/pkg/storeuploader/storeuploader.go]: I called it `keep`.

**Pull request merged**